### PR TITLE
feat(richeditor): native @tiptap/markdown AST on Tiptap 3.26.1

### PR DESCRIPTION
**Pull request by DeJeune**
### What this PR does

Merges the two foundational Tiptap PRs (upgrade + native Markdown) into one change for the Notes **RichEditor**.

Before this PR:
- Editor pinned to **Tiptap 3.2.0**.
- Markdown was handled by an HTML bridge in `utils/markdownConverter` (`markdown-it` → HTML → ProseMirror on parse, `turndown` on serialize), with bespoke `markdown-it`/`turndown` plugins for YAML front matter, math and task lists.

After this PR:
- All `@tiptap/*` packages upgraded to **3.26.1** and `@tiptap/markdown` added.
- The editor parses/serializes Markdown via the **native `@tiptap/markdown` (marked) AST**: `Markdown.configure({ markedOptions: { gfm: true } })`, `content` + `contentType: 'markdown'`, `editor.getMarkdown()`, and clipboard paste through the Markdown AST.
- Only **YamlFrontMatter** needed a custom `markdownTokenizer` + `parseMarkdown`/`renderMarkdown` (it is the one block standard Markdown doesn't cover). Math, code blocks, lists/tasks, images, links and underline inherit native Markdown support from their base 3.26 extensions.
- Upgrade fixups: `pnpm.overrides` pin `prosemirror-model`/`-view`/`-transform` to single instances (avoids duplicate-instance schema mismatches); `placeholder.ts` imports the PM `Node` type from `@tiptap/pm/model`; `EnhancedImage` casts to `ImageOptions` (3.26 made `inline`/`resize` required); the drag-handle scroll-hide pnpm patch was regenerated for 3.26.1; `@cherrystudio/extension-table-plus` dev `@tiptap` deps bumped to 3.26.1.
- `markdownConverter` is retained only for editor-less preview text and the auxiliary `html`/`setHtml` hook API.

Fixes #N/A

### Why we need it and why it was done in this way

Native Markdown is the prerequisite for the planned AI editing features (read/edit-the-document, diff, suggestions) and removes the lossy HTML round-trip.

The following tradeoffs were made:
- Kept `markdownConverter` for note-list preview text (no editor instance available there) rather than rewriting that path in the same PR.
- Pinned ProseMirror packages via `pnpm.overrides` (matching the repo's existing `@codemirror/*` approach) instead of per-package version juggling.

The following alternatives were considered:
- **Rebasing the old `DeJeune/tiptap-markdown-migration` branch** (1469 commits behind `main`, on the pre-v2 `src/renderer/src/...` layout) — rejected as a catastrophic-conflict rebase; the work was **ported forward** onto `origin/main` instead.
- **Tiptap Pro AI Toolkit** for the later AI work — deferred; it is a paid private-registry add-on, and Cherry Studio's existing `aiCore` (AI SDK v5 + tools) can drive a custom ProseMirror layer.

Links to places where the discussion took place: N/A

### Breaking changes

None for end users — Notes content stays as Markdown on disk and round-trips were verified.
For contributors: the ProseMirror packages are now resolved through `pnpm.overrides`, and PM types must be imported from `@tiptap/pm/*` (not `prosemirror-*`) to keep a single instance.

### Special notes for your reviewer

- Commits are currently **unsigned** — this environment's 1Password SSH signing agent can't prompt headlessly. Will re-sign if branch protection requires it.
- Local verification: `pnpm typecheck:web` clean and the **full renderer suite (364 files, 4053 passed, 3 skipped)** green; new headless round-trip tests added via `MarkdownManager`. The complete `pnpm lint`/`pnpm test` gate is left to CI.
- The only custom Markdown hook added is `YamlFrontMatter`; please sanity-check table / nested-task-list round-trips, which now flow through marked (v17) rather than the old `turndown` rules.

### Checklist

- [x] Branch: This PR targets the correct branch — `main` for active development, `v1` for v1 maintenance fixes
- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [x] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [x] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Documentation: A [user-guide update](https://docs.cherry-ai.com) was considered and is present (link) or not required. Check this only when the PR introduces or changes a user-facing feature or behavior.
- [x] Self-review: I have reviewed my own code (e.g., via [`/gh-pr-review`](/.claude/skills/gh-pr-review/SKILL.md), `gh pr diff`, or GitHub UI) before requesting review from others

### Release note

```release-note
NONE
```

### CODE REVIEW

File: src/renderer/components/RichEditor/useRichEditor.ts [hunk 1]
```diff
@@ -418,21 +292,22 @@ export const useRichEditor = (options: UseRichEditorOptions = {}): UseRichEditor
...
-          const html = markdownToHtml(text)
           const { $from } = selection
           const atStartOfLine = $from.parentOffset === 0
           const inEmptyParagraph = $from.parent.type.name === 'paragraph' && $from.parent.textContent === ''
+          const hasMultipleLines = text.includes('\n')
 
-          if (!atStartOfLine && !inEmptyParagraph) {
-            const cleanHtml = html.replace(/^<p>(.*?)<\/p>/s, '$1')
-            editor.commands.insertContent(cleanHtml)
+          if (!atStartOfLine && !inEmptyParagraph && !hasMultipleLines) {
```
Comment (AtomsH4) [inline src/renderer/components/RichEditor/useRichEditor.ts:304]: [Warning] 这里会让非空段落内粘贴的单行 Markdown 走 `insertText`，因此 `[docs](https://x)` / `**bold**` 会作为普通文本进入文档；保存时 `editor.getMarkdown()` 会再把这些 Markdown 标记转义，和迁移前 `markdownToHtml -> insertContent` 的行为不一致。

最小修复：只对 code block/明确需要 literal 的场景用 `insertText`，普通 inline Markdown 可解析成 paragraph 的 inline content 后插入；补一个非空段落内粘贴 link/bold 后保存的回归测试。

File: src/renderer/components/RichEditor/helpers/jumpToLine.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  totalLines?: number
+): HTMLElement | null {
+  const blocks = Array.from(editorDom.children).filter((el): el is HTMLElement => el instanceof HTMLElement)
+  if (blocks.length === 0) {
+    logger.warn('No editor blocks found for jump-to-line')
+    return null
+  }
+
+  // Strategy 1: content match against the rendered text of a top-level block.
+  const needle = lineContent ? normalizeMarkdownLine(lineContent) : ''
```
Comment (AtomsH4) [inline src/renderer/components/RichEditor/helpers/jumpToLine.ts:60]: [Warning] 这里会在重复文本时总是命中第一个 block，`lineNumber` 只在完全没有文本匹配时才参与比例估算。Notes 搜索结果的 `LOCATE_NOTE_LINE` 会传真实 `lineNumber + lineContent`，所以同一段文字出现多次时点击后可能高亮错误位置；表格分隔行等无可见文本也会靠比例跳错。

最小修复：恢复/重建 source-line 到 rendered block 的映射，或至少用 `lineNumber` 缩小候选范围后再匹配 `lineContent`；补重复文本和表格行的跳转测试。

File: src/renderer/components/RichEditor/extensions/yamlFrontMatter.ts [hunk 3]
```diff
@@ -17,6 +17,51 @@ export const YamlFrontMatter = Node.create({
...
+  },
+
+  renderMarkdown(node) {
+    const content = node.attrs?.content || ''
+    if (!content.trim()) {
+      return ''
+    }
+    // The stored content has no delimiters; add them back. Tolerate the legacy case
+    // where a trailing `---` was kept inside the content attribute.
+    if (content.endsWith('---')) {
```
Comment (AtomsH4) [inline src/renderer/components/RichEditor/extensions/yamlFrontMatter.ts]: [Warning] `content.endsWith("---")` 会把正常 YAML 值误判成 legacy closing fence。例如 front matter body 为 `title: foo---` 时会走这里，输出缺少单独的 closing `---`，后面的正文边界会被改变。

最小修复：只把 `/\n---\s*$/` 这类独立 closing delimiter 当 legacy case；或者内部统一存无 delimiter 内容，render 时总是补一行 closing fence。

**Review by AtomsH4 — COMMENTED** [review]
去重后补充 3 个 inline review comments。

File: src/renderer/components/RichEditor/extensions/yamlFrontMatter.ts (hunk 3 above)
Comment (eeee0717) [reply src/renderer/components/RichEditor/extensions/yamlFrontMatter.ts]: same comment

File: src/renderer/components/RichEditor/extensions/yamlFrontMatter.ts [hunk 4]
```diff
@@ -17,6 +17,51 @@ export const YamlFrontMatter = Node.create({
...
 
+  // Native markdown parse/serialize. YAML front matter (`---` … `---`) is not part of
+  // standard markdown, so it needs its own marked tokenizer to be picked up by @tiptap/markdown.
+  markdownTokenizer: {
+    name: 'yamlFrontMatter',
+    level: 'block',
+    start(src: string) {
+      return src.match(/^---\n/) ? 0 : -1
+    },
+    tokenize(src: string, tokens: MarkdownToken[] = []) {
```
Comment (zhangjiadi225) [inline src/renderer/components/RichEditor/extensions/yamlFrontMatter.ts]: **[A1]** The tokenizer only checks the current Marked token array to enforce “front matter must be at the top”. Marked re-lexes blockquote contents with a fresh token array, so a quoted block like:

```md
> ---
> title: quoted
> ---
```

can be converted into a nested `yamlFrontMatter` node. That changes ordinary quoted Markdown into a front-matter node inside the document. Please gate this on the absolute root parse context, or pre-extract only root front matter before Markdown parsing, and add a blockquote regression test.

File: src/renderer/components/RichEditor/extensions/markdownTable.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+            .renderChildren(cellNode.content || [])
+            .replace(/\r?\n+/g, ' ')
+            .replace(/\|/g, '\\|')
+            .trim()
+        )
+        .join(' | ')} |`
+
+    const columns = (rows[0].content || []).length
+    const separator = `| ${Array.from({ length: columns }, () => '---').join(' | ')} |`
+    return `${[renderRow(rows[0]), separator, ...rows.slice(1).map(renderRow)].join('\n')}\n\n`
```
Comment (zhangjiadi225) [inline src/renderer/components/RichEditor/extensions/markdownTable.ts]: **[A1]** This serializer always treats `rows[0]` as the Markdown header row. A table can lose its header row through the row action menu, for example after inserting the default header table and deleting the first row, leaving a first row made of `tableCell`s. On save/reload, this emits that body row before the separator, so it comes back as header cells and changes table semantics. Please detect whether the first row actually contains header cells; for headerless tables, emit a safe empty header/separator fallback and keep all body rows as body rows. Add a round-trip test for a table whose first row contains `tableCell`s only.

File: src/renderer/components/RichEditor/useRichEditor.ts [hunk 6]
```diff
@@ -343,53 +247,23 @@ export const useRichEditor = (options: UseRichEditorOptions = {}): UseRichEditor
...
     // eslint-disable-next-line react-hooks/exhaustive-deps
     [placeholder, activeShikiTheme, handleLinkHover, handleLinkHoverEnd]
   )
 
   const editor = useEditor({
     shouldRerenderOnTransaction: true,
     extensions,
-    content: html || '',
+    content: markdown || '',
+    contentType: 'markdown',
```
Comment (zhangjiadi225) [inline src/renderer/components/RichEditor/useRichEditor.ts:267]: **[C6]** This path now always parses editor input as Markdown, but the public `RichEditorProps` contract still says `initialContent` can be Markdown or HTML and still exposes `isMarkdown?: boolean`; `SkillDetailPage` also still passes `isMarkdown={true}`. That leaves callers with an API that appears to support HTML/auto-detection even though this implementation no longer does. Please either remove the stale `isMarkdown` prop and update the `initialContent` documentation/call sites to Markdown-only, or restore an explicit HTML compatibility path.

File: package.json [hunk 7]
```diff
@@ -216,34 +216,33 @@
...
+    "@tiptap/extension-node-range": "3.26.1",
+    "@tiptap/extension-table-of-contents": "3.26.1",
+    "@tiptap/extension-typography": "3.26.1",
+    "@tiptap/extension-underline": "3.26.1",
+    "@tiptap/markdown": "3.26.1",
+    "@tiptap/pm": "3.26.1",
+    "@tiptap/react": "3.26.1",
+    "@tiptap/starter-kit": "3.26.1",
+    "@tiptap/suggestion": "3.26.1",
     "@tiptap/y-tiptap": "^3.0.0",
```
Comment (zhangjiadi225) [inline package.json]: **[B4]** The rest of the Tiptap stack was moved to `3.26.1`, but `@tiptap/y-tiptap` still allows/resolves to `3.0.1` in the lockfile. The new `@tiptap/extension-collaboration@3.26.1` and `@tiptap/extension-drag-handle@3.26.1` entries both peer-require `@tiptap/y-tiptap: ^3.0.5`, so the current lockfile keeps a real peer mismatch in the upgraded editor stack. Please bump this dependency to `^3.0.5` or a pinned compatible version and refresh `pnpm-lock.yaml`.

**Review by zhangjiadi225 — CHANGES_REQUESTED** [review]
I reviewed the native Markdown migration on a dedicated worktree at PR head `05cdc71cac2e3384a09c5dac40a4ef2f41aec705` and cross-checked the findings with a verifier pass. I am requesting changes because several confirmed issues can alter saved note content, break search navigation, or leave the upgraded Tiptap dependency graph inconsistent.

Confirmed issues:

1. **[A1]** `yamlFrontMatter` can still be parsed inside blockquotes. The tokenizer only checks the current Marked token list, but blockquote contents are lexed with their own token list, so quoted `--- ... ---` blocks can become nested front-matter nodes.

2. **[A1]** Front matter content that legitimately ends with `---` can serialize without a closing delimiter. This is already covered by the existing inline comment on `yamlFrontMatter.ts:59`, so I did not duplicate it.

3. **[A1]** Headerless tables can round-trip with the first body row promoted into Markdown header cells. I left an inline comment on `markdownTable.ts:58`.

4. **[A1]** RichEditor jump-to-line can scroll to the wrong block because content matching returns the first top-level block and ignores `lineNumber` once text matches. This is already covered by the existing inline comment on `jumpToLine.ts`, so I did not duplicate it.

5. **[B4]** `@tiptap/y-tiptap` still resolves to `3.0.1`, while the upgraded Tiptap collaboration/drag-handle packages peer-require `^3.0.5`. I left an inline comment on `package.json:239`.

6. **[C6]** The public `RichEditor` contract still advertises HTML/auto-detect semantics via `initialContent` comments and `isMarkdown`, while the implementation now always parses content as Markdown. I left an inline comment on the new `contentType: 'markdown'` path.

CI is green, but these edge cases are not covered by the current tests. The smallest follow-up is to add targeted round-trip/search tests for the affected Markdown constructs and update the stale public API contract.

File: src/renderer/components/RichEditor/helpers/jumpToLine.ts (hunk 2 above)
Comment (DeJeune) [reply src/renderer/components/RichEditor/helpers/jumpToLine.ts:60]: 确认问题成立,已在 `cc89f1d965` 修复。

`findElementByLine` 现在先用 `lineNumber / totalLines` 估算目标块位置:文本在多个块里重复命中时,取离估算位置最近的那个块,而不是固定返回第一个;只有完全没有文本匹配(表格分隔行、fence 等无可见文本)时才退回纯比例定位。已补一条重复文本的回归测试。

受限于 marked 不暴露 per-node 行号,rich view 仍是 best-effort(块级精度);source/code view 走 CodeMirror 真实行号,定位依旧精确。

File: src/renderer/components/RichEditor/extensions/yamlFrontMatter.ts (hunk 3 above)
Comment (DeJeune) [reply src/renderer/components/RichEditor/extensions/yamlFrontMatter.ts]: 确认是 bug,已在 `a88dd99720` 修复。

把判断从 `content.endsWith('---')` 收紧为只匹配独立成行的 closing delimiter `/\n---\s*$/`。这样 `title: foo---` 这类「值本身以 --- 结尾」的 front matter 仍会补上单独的 closing fence;而 content 里确实已带末尾 `---` 行的 legacy 数据继续兼容。已补两条回归测试:误判值 + legacy 容错。

File: src/renderer/components/RichEditor/extensions/markdownTable.ts (hunk 5 above)
Comment (DeJeune) [reply src/renderer/components/RichEditor/extensions/markdownTable.ts]: Fixed in `d625689e5f`. `renderMarkdown` now only treats the first row as the GFM header when it actually contains `tableHeader` cells. For a headerless table (e.g. after deleting the header row via the row action menu, leaving a `tableCell`-only first row) it synthesizes an empty header and keeps every row as a body row, so a body row is no longer promoted to a header on reload. Added round-trip tests for both the real-header and headerless cases.

File: src/renderer/components/RichEditor/useRichEditor.ts (hunk 6 above)
Comment (DeJeune) [reply src/renderer/components/RichEditor/useRichEditor.ts:267]: Fixed in `90d67c175f`. Removed the stale `isMarkdown` prop from `RichEditorProps`, dropped `isMarkdown={true}` at the `SkillDetailPage` call site, and reworded the `initialContent` doc to markdown-only. We did not restore an HTML path — Markdown via the native AST is the single source of truth now — so the public contract matches the implementation.

File: package.json (hunk 7 above)
Comment (DeJeune) [reply package.json]: Fixed in `571cdb0990`. Bumped `@tiptap/y-tiptap` to `^3.0.5` and refreshed `pnpm-lock.yaml` (now resolves `3.0.5`), matching the `^3.0.5` peer required by `@tiptap/extension-collaboration@3.26.1` and `@tiptap/extension-drag-handle@3.26.1`.

File: src/renderer/components/RichEditor/extensions/yamlFrontMatter.ts (hunk 4 above)
Comment (DeJeune) [reply src/renderer/components/RichEditor/extensions/yamlFrontMatter.ts]: Confirmed the repro — `> ---\n> title: quoted\n> ---` parses into a nested `yamlFrontMatter` node inside the blockquote.

Root cause: the tokenizer's "must be at the top" guard inspects only the current Marked token array, and Marked re-lexes blockquote content with a fresh `[]`, so a root parse and a blockquote-inner parse look identical to the guard.

A clean in-tokenizer gate isn't directly available: `@tiptap/markdown` invokes the extension as `tokenize(src, tokens, helper)` without the Marked `Lexer`, and the `helper` only exposes `inlineTokens`/`blockTokens`, so the tokenizer can't read `lexer.state.top` (root vs. nested). So the fix is one of: (a) pre-extract only the leading root front matter before Markdown parsing and re-attach the delimiters on serialize, or (b) a small `@tiptap/markdown` patch to forward `lexer.state.top` to the tokenizer callback (the repo already patches `@tiptap/extension-drag-handle`, so a patch is established here). I'll send this as a focused follow-up with a blockquote regression test.

Comment (DeJeune) [reply src/renderer/components/RichEditor/extensions/yamlFrontMatter.ts]: Implemented in `16d5239ff9`.

One correction to my earlier note: `lexer.state.top` is `true` inside blockquotes too (marked only sets it `false` for list items), so it can't tell a root parse from a blockquote-inner parse. The reliable "absolute root" signal is identity against the lexer's root token array — `tokens === lexer.tokens` is `true` only at the document root (`false` in blockquotes, lists, and nested blockquotes).

So I went with approach (b): a small `@tiptap/markdown` patch (`patches/@tiptap__markdown@3.26.1.patch`) that surfaces the marked `lexer` on the tokenizer helper bag, and the front-matter tokenizer now bails out when `tokens !== lexer.tokens`. Added a regression test asserting `> --- … ---` stays ordinary blockquote content instead of becoming a nested front-matter node.

File: src/renderer/components/RichEditor/useRichEditor.ts (hunk 1 above)
Comment (DeJeune) [reply src/renderer/components/RichEditor/useRichEditor.ts:304]: 已在 `a9dd271087` 修复。

非空段落内粘贴单行 Markdown 时,现在先用 `editor.markdown.parse(text)` 解析,只取「单段落的 inline content」插入——这样 `**bold**` / `[text](url)` 会变成真正的 mark 而不是字面文本(保存时不再被 `getMarkdown` 转义成 `\*\*bold\*\*`),同时不会把粘贴内容包成新 block。若该行是 heading / list 等块级语法(没有纯 inline 形态),则回退到原来的 literal 文本插入。

补了回归测试(`helpers/__tests__/markdownPaste.test.ts`):在非空段落 `Hello ` 内粘贴 `**bold** and [docs](https://x.com)`,`getMarkdown()` 应得到 `Hello **bold** and [docs](https://x.com)`,markers 不被转义。

**Review by AtomsH4 — APPROVED** [review]
已复核最新改动，除已确认忽略的 paste inline Markdown warning 外，前序问题已修复。Approve。

**Review by eeee0717 — APPROVED** [review]

**Review by zhangjiadi225 — APPROVED** [review]
复审通过 ✅ — 此前 request changes 的 6 个问题已在 worktree 上逐条核实修复,且都带了回归测试,CI 全绿。从 changes requested 改为 approve。

## 总评

这是一次质量很高的基础设施迁移:把 Notes RichEditor 从「markdown-it/turndown 的 HTML 桥接」切换到 `@tiptap/markdown` 原生 marked AST,并将整个 Tiptap 栈升到 3.26.1。改动聚焦、删多于增(去掉有损的 HTML round-trip),为后续 AI 编辑能力打好了地基。难点处理得当:

- 只有 YamlFrontMatter 这一标准 Markdown 不覆盖的块需要自定义 tokenizer,其余(数学、代码、列表/任务、图片、链接、下划线)都复用 3.26 基础扩展的原生能力,克制。
- ProseMirror 单实例通过 `pnpm.overrides` 收敛、PM 类型从 `@tiptap/pm/*` 引入,避免重复实例导致的 schema 不一致——是正确做法。
- 升级链(drag-handle 补丁重生成、y-tiptap peer、extension-table-plus dev deps)都对齐到位。

边界用例此前是主要风险点(会改动已保存内容、影响搜索跳转),现已全部覆盖并加测。

## 上一轮 6 个问题逐条确认

1. **[A1] blockquote 内 front matter**(`16d5239ff`):marked 补丁透出 `lexer`,用根 token 数组同一性 `tokens !== lexer.tokens` 判断绝对根(比 `lexer.state.top` 可靠,blockquote 里 top 也是 true);补丁缺失时优雅回退。`> --- … ---` 回归测试已加。
2. **[A1] front matter 以 `---` 结尾丢失 closing fence**(`a88dd9972`):收紧为独立成行的 `/\n---\s*$/`,legacy 兼容保留,带两条测试。
3. **[A1] 无表头表格首行被提升为表头**(`d625689e5`):仅当首行含 `tableHeader` 才当表头,否则合成空表头并保留全部 body 行,带 round-trip 测试。
4. **[A1] jump-to-line 重复文本误命中**(`cc89f1d96`):用 `lineNumber/totalLines` 估算位置后就近消歧,调用方确实计算并传入 `totalLines`,非空操作;带重复文本测试。受 marked 无 per-node 行号限制,rich view 为块级 best-effort、source/code view 走 CodeMirror 精确行号,合理。
5. **[B4] `@tiptap/y-tiptap` peer 不匹配**(`571cdb099`):升到 `^3.0.5`,lockfile 全量解析 `3.0.5`,满足 collaboration/drag-handle 的 `^3.0.5` peer。
6. **[C6] 公共契约残留 HTML/`isMarkdown`**(`90d67c175`):删除 `isMarkdown` prop、`SkillDetailPage` 调用点同步、`initialContent` 文档改为 markdown-only,契约与实现一致。

补丁已正确登记在 `pnpm.patchedDependencies`;AtomsH4 的粘贴 inline Markdown 与 jumpToLine 两条 Warning 也已修(`a9dd27108`/`cc89f1d96`)。

LGTM,可合并。
